storage_service/locations/models/s3.py

```diff
@@ -21,68 +22,114 @@
 LOGGER = logging.getLogger(__name__)
 
 
-class S3(models.Model):
-    space = models.OneToOneField("Space", to_field="uuid")
-    endpoint_url = models.CharField(
-        max_length=2048,
-        verbose_name=_("S3 Endpoint URL"),
-        help_text=_("S3 Endpoint URL. Eg. https://s3.amazonaws.com"),
+class S3SpaceModelMixin(models.Model):
+
+    class Meta:
+        app_label = 'locations'
+        abstract = True
+
+    # Authentication details
+    aws_access_key_id = models.CharField(
+        max_length=64,
+        blank=True,
+        verbose_name=_('Access Key ID to authenticate')
     )
-    access_key_id = models.CharField(
-        max_length=64, verbose_name=_("Access Key ID to authenticate")
+
+    aws_secret_access_key = models.CharField(
+        max_length=256,
+        blank=True,
+        verbose_name=_('Secret Access Key to authenticate with')
     )
-    secret_access_key = models.CharField(
-        max_length=256, verbose_name=_("Secret Access Key to authenticate with")
+
+    aws_assumed_role = models.CharField(
+        max_length=256,
+        blank=True,
+        verbose_name=_('Assumed AWS IAM Role')
     )
-    region = models.CharField(
-        max_length=64,
-        verbose_name=_("Region"),
-        help_text=_("Region in S3. Eg. us-east-2"),
+
+    s3_endpoint_url = models.CharField(
+        max_length=2048,
+        verbose_name=_('S3 Endpoint URL'),
+        help_text=_('S3 Endpoint URL. Eg. https://s3.amazonaws.com')
     )
 
-    class Meta:
-        verbose_name = _("S3")
-        app_label = "locations"
+    s3_region = models.CharField(
+        max_length=64,
+        verbose_name=_('Region'),
+        help_text=_('Region in S3. Eg. us-east-2')
+    )
 
-    ALLOWED_LOCATION_PURPOSE = [Location.AIP_STORAGE]
+    s3_bucket = models.CharField(max_length=64,
+        verbose_name=_('S3 Bucket'),
+        blank=True,
+        help_text=_('S3 Bucket Name'))
 
-    def __init__(self, *args, **kwargs):
-        super(S3, self).__init__(*args, **kwargs)
-        self._client = None
-        self._resource = None
+    def _ensure_bucket_exists(self):
+        client = self.s3_resource.meta.client
+        try:
+            client.head_bucket(Bucket=self.bucket_name)
+        except botocore.exceptions.ClientError as e:
+            # If a client error is thrown, then check that it was a 404 error.
+            # If it was a 404 error, then the bucket does not exist.
+            error_code = e.response['Error']['Code']
+            if error_code == '404':
+                client.create_bucket(Bucket=self.bucket_name)
```

Comment:
What if HeadBucket fails and it’s not a 404? What if CreateBucket fails? We should log/throw those exceptions.

Although the pattern we’ve adopted elsewhere is that applications get passed infrastructure as parameters, and aren’t responsible for creating it themselves – this would create a bucket for Archivematica if it didn't exist, but couldn't assign permissions to itself to read it. I think we could do without this check – if it tries to write to a non-existent bucket, just fail and throw an exception.

Reply:
The ensure_bucket_exists functionality is already present in the s3 plugin so I'm inclined to leave it in, although I'm not sure how well it worked or how used it was, given it would previously have raised an exception if the bucket already existed.